File: lib/bodies/product.dart
import 'package:mango_ui/keys.dart';

class Product {
  final Key categoryKey;
  final List<Key> itemKeys;
  final String shortname;
  final Key imageKey;
  final DateTime expires;
  final String currency;
  final double price;
  final double estimate;
  final List<String> tags;
  final String location;
  final int views;

  Product(
      this.categoryKey,
      this.itemKeys,
      this.shortname,
      this.imageKey,
      this.expires,
      this.currency,
      this.price,
      this.estimate,
      this.tags,
      this.location,
      this.views);

  Map<String, dynamic> toJson() {
    return {
      "CategoryKey": categoryKey,
      "ItemKeys": itemKeys,
      "ImageKey": imageKey,
      "Expires": expires.toIso8601String(),
      "Price": price,
      "Tags": tags,
      "Location": location,
      "Views": views,
      "EstimateValue": estimate,
      "Currency": currency,
      "ShortName": shortname,
    };
  }
}

File: core/product.go
package core

import (
	"github.com/louisevanderlith/husk/hsk"
	"github.com/louisevanderlith/husk/validation"
	"time"
)

type Product struct {
	CategoryKey   hsk.Key
	ItemKeys      []hsk.Key
	ShortName     string `hsk:"size(128)"`
	ImageKey      hsk.Key
	Expires       time.Time
	Currency      string
	Price         float32 //User set Value
	EstimateValue float32 //Total of all Products
	Tags          []string
	Location      string `hsk:"size(128)"`
	Views         int64  `hsk:"null"`
}

func (s Product) Valid() error {
	err := validation.Struct(s)

	if err != nil {
		return err
	}

	return PriceInBounds(s.Price, s.EstimateValue)
}
